civiform/civiform
# Add Universal question tag and toggle

### Description

- Our current toggle reloads the page (or, rather, part of the page via htmx) every time it is toggled because we need to record the state of the toggle in the program block edit view, which doesn't have a "submit" kind of button. Here, we want the toggle to behave more like a checkbox. This adds a toggle that is defined much like our current toggle, but sets the state of a hidden checkbox underneath and is toggled via typescript.  We should be able to rework our current toggle to use this new ViewUtils.makeToggleButton later on.  This is currently styled exactly the same as our current toggle, which doesn't quite match the mocks.  There may be some tweaks pending final approval from @sijiayam . Note that this PR originally contained a brand new toggle component written mostly in CSS.  The commit remains in the PR history, but reverted, in case we decide we want to go that direction instead.  However, making the toggle a button is better from an accessibility standpoint.  
- Adds a toggle to the question create/edit view that sets whether the question is a universal question.
- Adds a `isUniversal` field to `QuestionForm`, which binds to the state of the toggle used in the view.
- Adds a new `QuestionTag` called `UNIVERSAL`.  When present for a question, that question will be treated as a universal question, with changes in the UI that will be added in later PRs.
- Adds a `universal` property and `setUniversal` function to QuestionDefinitionConfig, a `setUniversal` function to `QuestionDefinitionBuilder`, and a `isUniversal` function to `QuestionDefinition`.  The `universal` property will be default `false`, set via the config builder.
- When a `Question` is created by passing in a `QuestionDefinition` to the constructor, the `UNIVERSAL` QuestionTag will get added if `QuestionDefinition.isUniversal()` is true.
- Adds a UNIVERSAL_QUESTIONS feature flag.
- Optimizes the code in some of the unit tests

### Checklist

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)


#### User visible changes

Note: I did not leave the validateAccessibility function in the test, since it fails for other things on the page, but verified there are no warnings for the added components.

#### New Features

- [x] Add new FeatureFlag env vars to `server/conf/helper/feature-flags.conf`

### Instructions for manual testing

1. Turn on the feature flag in the settings UI
2. Create a new question with universal set/unset, and ensure it remains set/unset when editing the question.
3. Edit the question, toggling universal, and ensure the state persists.

### Issue(s) this completes

https://github.com/civiform/civiform/issues/5860 https://github.com/civiform/civiform/issues/5810
